src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java

```diff
@@ -0,0 +1,127 @@
+import com.google.common.collect.Lists;
+import htsjdk.samtools.*;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.broadinstitute.hellbender.engine.ReadsDataSource;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.GoogleGenomicsReadToGATKReadAdapter;
+import org.seqdoop.hadoop_bam.AnySAMInputFormat;
+import org.seqdoop.hadoop_bam.SAMRecordWritable;
+import org.seqdoop.hadoop_bam.util.SAMHeaderReader;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+
+/** Loads the reads from disk either serially (using samReaderFactory) or in parallel using Hadoop-BAM.
+ * The parallel code is a modified version of the example writing code from Hadoop-BAM.
+ */
+public class ReadsSparkSource {
+    private final JavaSparkContext ctx;
+    public ReadsSparkSource(JavaSparkContext ctx) {
+        this.ctx = ctx;
+    }
+
+    /**
+     * Loads Reads using samReaderFactory, then calling ctx.parallelize.
+     * @param bam file to load
+     * @return RDD of (SAMRecord-backed) GATKReads from the file.
+     */
+    public JavaRDD<GATKRead> getSerialReads(final String bam) {
+        final SAMFileHeader readsHeader = getHeader(bam);
+        List<SimpleInterval> intervals = IntervalUtils.getAllIntervalsForReference(readsHeader.getSequenceDictionary());
+        final SamReaderFactory samReaderFactory = SamReaderFactory.makeDefault().validationStringency(ValidationStringency.SILENT);
+
+        ReadsDataSource bam2 = new ReadsDataSource(new File(bam), samReaderFactory);
+        bam2.setIntervalsForTraversal(intervals);
+        List<GATKRead> records = Lists.newArrayList();
+        for ( GATKRead read : bam2 ) {
+            records.add(read);
+        }
+        return ctx.parallelize(records);
+    }
+
+    /**
+     * Loads Reads using Hadoop-BAM. For local files, bam must have the fully-qualified path,
+     * i.e., file:///path/to/bam.bam.
+     * @param bam file to load
+     * @param intervals intervals of reads to include.
+     * @return RDD of (Google Read-backed) GATKReads from the file.
+     */
+    public JavaRDD<GATKRead> getParallelReads(final String bam, final List<SimpleInterval> intervals) {
+        JavaPairRDD<LongWritable, SAMRecordWritable> rdd2 = ctx.newAPIHadoopFile(
+                bam, AnySAMInputFormat.class, LongWritable.class, SAMRecordWritable.class,
+                new Configuration());
+
+        return rdd2.map(v1 -> {
+            SAMRecord sam = v1._2().get();
+            if (samRecordOverlaps(sam, intervals)) {
```

Comment:
Does Hadoop-BAM natively support query by intervals? Checking every read against all intervals could get costly if the interval list is very large (as is the case when working with exome data, typically)